diff-containers/README.md

```diff
@@ -1 +1,238 @@
+   fromList [(1, Insert_ 54), (2, DDelete 42)                ]
+<> fromList [                 (2, Insert_ 17), (3, DDelete 1)]
+== fromList [(1, Insert_ 54), (2, Insert_ 17), (3, DDelete 1)]
+```
+
+What if we want to remove diffs, i.e. through subtraction provided by a `Group`
+instance? In the current definition, there is no way to recover information
+about previous changes, since we only have the last/most recent change. In
+particular, we would like to be able to have the following:
+
+```haskell
+   fromList [(1, Insert_ 54), (2, DDelete 42)                ]
+<> fromList [                 (2, Insert_ 17), (3, DDelete 1)]
+~~ fromList [                 (2, Insert_ 17), (3, DDelete 1)]
+== fromList [(1, Insert_ 54), (2, DDelete 42)                ]
+```
+
+The current definition of diffs is not sufficient, however.
+
+### 3) Diffs with inverses
+
+The trick is to store a *history* of deltas for each key. When we subtract diffs
+from one another, we remove deltas from histories. How exactly deltas are
+removed, we will touch upon later. For now, these are the (simplified) types
+that make up a diff:
+
+```haskell
+-- Note: this is a simplified representation. In reality, we enforce statically
+-- that diff histories in a @'Diff'@ are non-empty.
+newtype Diff k v = Diff (Map k (DiffHistory v))
+
+newtype DiffHistory v = DiffHistory { getDiffHistory :: Seq (DiffEntry v) }
+
+data DiffEntry v =
+    Insert !v
+  | Delete !v
+  {- omitted for now -}
+```
+
+The way in which subtraction is facilitated is by implementing the `invert`
+method in a `Group` instance. As a consequence, not only do we have to add
+inverses for diffs, diff histories and diff entries, we must also define how
+these inverses are handled by summing.
+
+First, we start by adding inverse elements to the definition of `DiffEntry`. One
+may think `Insert` and `Delete` are already inverses of one another, and that
+would be correct in terms of adhering to the type class laws, but it would have
+adverse effects on the semantics of applying diffs to maps. Consider what would
+happen if we were to consider `Insert` and `Delete` inverses in the following
+example. Ignore diff histories for now:
+
+```haskell
+d1 = fromList [(0, Delete 17)]
+d2 = fromList [(0, Insert 42)]
+d3 = d1 <> d2 = fromList []
+```
+
+Applying `d3` to a map would behave like the identity function. Applying `d1`
+and `d2` consecutively would not behave like the identity function, as it would
+map `0` to `42`. Through some abuse of the definition of distributivity, we can
+say that applying diffs does not distribute over the summing of diffs.
+
+Instead, we add explicit inverse constructors for both `Insert` and `Delete`. We
+do not make `Semigroup`/`Monoid`/`Group` instances for diff entries, however. We
+shall see that we only need these 4 constructors in order to provide the
+instances for diff histories.
+
+```haskell
+data DiffEntry v =
+    Insert !v
+  | Delete !v
+  | UnsafeAntiInsert !v -- ^ Unsafe!
```

Comment:
Maybe we should explain why these are unsafe.

Reply:
Done